File: bridgeAddon/tests/fakes/listener.py
# nvdaMcpBridge tests -- FakeListener, standing in for the Listener seam.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
#
# FAKES: adapters/ports/listener.py
#
# Lets a test drive BridgeServer's accept loop precisely, with no socket: queue
# a connection with connect(), let accept() report TimeoutError while idle
# (exactly as a real listener polls), arm a one-shot fault with fail_next_accept,
# and close() to make accept() raise ListenerClosed. It records open/close so a
# test can assert the server bound once and released the socket on stop.

from __future__ import annotations

import queue
import threading

from nvdaMcpBridge.adapters.ports.listener import Listener, ListenerClosed
from nvdaMcpBridge.adapters.ports.transport import Transport


class FakeListener(Listener):
	"""Yields scripted connections; times out when idle; ListenerClosed once closed."""

	def __init__(self, endpoint: str = "127.0.0.1:0", *, poll_timeout: float = 0.02) -> None:
		self._endpoint = endpoint
		self._poll_timeout = poll_timeout
		self._connections: queue.Queue[Transport] = queue.Queue()
		self._closed = threading.Event()
		self._fault: Exception | None = None
		self.opened = False
		self.closes = 0

	# -- test hooks ----------------------------------------------------------

	def connect(self, transport: Transport) -> None:
		"""Queue a connection the next accept() will return."""
		self._connections.put(transport)

	def fail_next_accept(self, error: Exception) -> None:
		"""Arm a one-shot fault: the next accept() raises ``error``."""
		self._fault = error

	# -- Listener ------------------------------------------------------------

	@property
	def endpoint(self) -> str:
		return self._endpoint

	def open(self) -> None:
		self.opened = True

	def accept(self) -> Transport:
		if self._closed.is_set():
			raise ListenerClosed
		if self._fault is not None:
			fault, self._fault = self._fault, None
			raise fault
		try:
			return self._connections.get(timeout=self._poll_timeout)
		except queue.Empty:
			if self._closed.is_set():
				raise ListenerClosed from None
			raise TimeoutError from None

	def close(self) -> None:
		self.closes += 1
		self._closed.set()

File: bridges/voiceover/Tests/Fakes/Listener.swift
// Hand-written stateful fake for the Listener seam, scripted with connections, idle polls and faults.

import Foundation
import VoiceOverBridgeAdapters

public final class FakeListener: Listener {
	public enum Step {
		case connection(any Transport)
		case idle
		case fault(any Error)
	}

	private let lock = NSLock()
	private var script: [Step]
	public let endpoint: String
	public private(set) var openCount = 0
	public private(set) var closeCount = 0

	public init(endpoint: String = "fake-endpoint", script: [Step] = []) {
		self.endpoint = endpoint
		self.script = script
	}

	public func open() throws {
		lock.lock()
		openCount += 1
		lock.unlock()
	}

	public func accept() throws -> any Transport {
		lock.lock()
		let closed = closeCount > 0
		let next = script.isEmpty ? nil : script.removeFirst()
		lock.unlock()
		if closed { throw ListenerClosed() }
		guard let next else { throw PollTimeout() }
		switch next {
		case .connection(let transport): return transport
		case .idle: throw PollTimeout()
		case .fault(let error): throw error
		}
	}

	public func close() {
		lock.lock()
		closeCount += 1
		lock.unlock()
	}
}
